### PR TITLE
Update to Unicode 9.0.0

### PR DESCRIPTION
**Pull request by LEOYoon-Tsaw**
#### Resolved bug number: ([SR-](https://bugs.swift.org/browse/SR-))

---

Before merging this pull request to apple/swift repository:
- [ ] Test pull request on Swift continuous integration.

<details>
  <summary>

Triggering Swift CI</summary>

The swift-ci is triggered by writing a comment on this PR addressed to the GitHub user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please smoke test |
| All supported platforms | @swift-ci Please smoke test and merge |
| OS X platform | @swift-ci Please smoke test OS X platform |
| Linux platform | @swift-ci Please smoke test Linux platform |

**Validation Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please test |
| All supported platforms | @swift-ci Please test and merge |
| OS X platform | @swift-ci Please test OS X platform |
| OS X platform | @swift-ci Please benchmark |
| Linux platform | @swift-ci Please test Linux platform |

**Lint Testing**

| Language | Comment |
| --- | --- |
| Python | @swift-ci Please Python lint |

Note: Only members of the Apple organization can trigger swift-ci.
</details>

### CODE REVIEW

**Comment by gribozavr** [comment]
Thank you!  Unfortunately, this is not sufficient -- you need to update a few more files, and it would be best to add a test.  Could you take a look at the previous update, https://github.com/apple/swift/commit/7ae303aeb31e44aaa4a6c8eea844282995b55587 and mirror what was done there?

**Comment by gribozavr** [comment]
Actually, this is not sufficient to update to Unicode 9.0, we will probably fail the conformance tests that are included in your change (would be good to check after you update the trie size; without updating the trie size anything accessing the trie should crash).

There were major updates to the segmentation algorithms -- http://www.unicode.org/reports/tr29/tr29-29.html#Modifications  We need to update the rules in `get_extended_grapheme_cluster_rules_matrix` in `utils/GYBUnicodeDataUtils.py`.

**Comment by LEOYoon-Tsaw** [comment]
I've made coresponding changes to UnicodeTrie.swift.gyb and GYBUnicodeDataUtils.py. But I'm not so sure about the TrieSize value. And since there are 5 more cluster properties added in Unicode 9.0.0, I tried my best to account for them, but if there's anything(very likely) left, please tell me.

File: utils/GYBUnicodeDataUtils.py [hunk 1]
```diff
@@ -63,6 +63,11 @@ class GraphemeClusterBreakPropertyTable(UnicodeProperty):
         'T': 10,
         'LV': 11,
         'LVT': 12,
+        'E_Base_GAZ': 13,
+        'E_Base': 14,
+        'E_Modifier': 15,
+        'Glue_After_Zwj': 16,
+        'ZWJ': 17
```
Comment (gribozavr) [inline utils/GYBUnicodeDataUtils.py]: Could you also update `include/swift/Basic/Unicode.h`?

**Comment by gribozavr** [comment]
@LEOYoon-Tsaw Seems like we are hitting an assertion failure:

```
  File "/home/buildnode/jenkins/workspace/swift-PR-Linux/swift/utils/GYBUnicodeDataUtils.py", line 525, in get_extended_grapheme_cluster_rules_matrix
    assert(len(any_value) <= 16)
AssertionError
```

There is an assumption that we can encode each row of the matrix that represents the rules in a `uint16_t`.  Now we have 18 property values, and this is no longer true.  We probably need to expand that to `uint32_t` and adjust the corresponding Swift and C++ code (`_UnicodeExtendedGraphemeClusterSegmenter`, `_swift_stdlib_ExtendedGraphemeClusterNoBoundaryRulesMatrixImpl`, `swift::unicode::ExtendedGraphemeClusterNoBoundaryRulesMatrix` etc.)

**Comment by LEOYoon-Tsaw** [comment]
Sorry I didn't notice that. But I'm afraid those further adjustments to make are less obvious. Maybe someone more familiar with the previous implimentations is needed.

**Comment by LEOYoon-Tsaw** [comment]
@gribozavr there is a new rule in 9.0.0, which is (E_Base | EBG) Extend* × E_Modifier.
I covered the case when there is no Entending character, but the current matrix structure seems unable to support the case when there are one or more Extending chars following the emoji base char.

**Comment by LEOYoon-Tsaw** [comment]
Done some upgration of Grapheme Cluster rules matrix size from 16bit to 32bit

**Comment by gribozavr** [comment]
@LEOYoon-Tsaw Thank you, this is great progress!

> @gribozavr there is a new rule in 9.0.0, which is (E_Base | EBG) Extend* × E_Modifier.

You are right about this.  The current algorithm can't support the new rules.  The current algorithm is stateless, it only looks at two adjacent Unicode scalars.  The new algorithm in Unicode 9 requires a state machine.

**Comment by LEOYoon-Tsaw** [comment]
@gribozavr In the current implementation, the matrix consists of 0 and 1s, representing break and not break. I'm considering replacing those bits with structures each consists of a Status Transition Matrix and a Boundary-Checking Vector.
Upon beginning processing a string, a Status Vector is assinged, by checking each adjacent Unicode scalar pairs, a bitwise-and between the Status Vector and the coresponding Boundary-Cheking Vector will determing whether a boundary is found, and then the status vector evolves by applying the coresponding Status Transition Matrix to it.
I'm not 100% sure this solution will best serve the new rules, I'm just posting my idea here and guessing if there is a better one. Cheers.

**Comment by CodaFi** [comment]
Superseded by the String overhaul, and in particular #9265.

**Pull request closed without merging**